File: core/src/main/java/in/koreatech/koin/core/util/ImageUtil.java
package in.koreatech.koin.core.util;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

public class ImageUtil {
    public static final String TAG = ImageUtil.class.getName();

    public static byte[] reduceSize(Bitmap bitmap, int maxSize) {
        int quality = 100;
        int redeceQuality = 2;
        while (true) {
            if (quality <= 0) return null;
            ByteArrayOutputStream bos = new ByteArrayOutputStream();
            Log.d(TAG, "quality: " + quality);
            bitmap.compress(Bitmap.CompressFormat.JPEG, quality, bos);
            byte[] bitmapdata = bos.toByteArray();
            Log.d(TAG, "size: " + bitmapdata.length/10000.0);
            if (bitmapdata.length /1000000.0 > maxSize) quality -= redeceQuality;
            else return bitmapdata;
        }
    }

    public static File changeBMPtoFILE(Bitmap bitmap, String fileName, int maxSize, Context context) throws IOException {
        File file = new File(context.getCacheDir(), fileName + ".jpg");
        file.createNewFile();


        byte[] bitmapdata = reduceSize(bitmap,maxSize);
        if(bitmapdata == null) return  null;
        Log.d(TAG, "size: " + bitmapdata.length/1000000.0 );
        FileOutputStream fos = null;
        try {
            fos = new FileOutputStream(file);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        try {
            fos.write(bitmapdata);
            fos.flush();
            fos.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return file;
    }
}

File: core/src/main/java/in/koreatech/koin/core/upload/ImageUtil.kt
package `in`.koreatech.koin.core.upload

import android.content.ContentValues
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ExifInterface
import android.net.Uri
import android.provider.MediaStore
import android.util.Log
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.util.Calendar


fun InputStream.toResizeBitmap(fileSize: Long): Bitmap? {

    return if(fileSize >= 1000000)
    {
        val option = BitmapFactory.Options()

        option.apply {
            inJustDecodeBounds = false
            inSampleSize = 4
        }

        BitmapFactory.decodeStream(this, null, option)
    }
    else
        BitmapFactory.decodeStream(this)
}

fun Bitmap.rotateBitmap(angle: Float): Bitmap {
    val matrix = Matrix()
    matrix.postRotate(angle)
    return Bitmap.createBitmap(this, 0, 0, this.width, this.height, matrix, true)
}

fun Bitmap.toCompressJPEG(): ByteArray {
        val byteArrayOutputStream = ByteArrayOutputStream()
        this.compress(Bitmap.CompressFormat.JPEG, 100, byteArrayOutputStream)

        return byteArrayOutputStream.toByteArray()
}
fun createImageFile(context: Context): Uri? {
    val contentResolver = context.contentResolver
    val content = ContentValues().apply {
        put(MediaStore.MediaColumns.MIME_TYPE, "image/jpeg")
    }
    return contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, content)
}